app/models/concerns/key_pair_compute_resource.rb

```diff
@@ -0,0 +1,55 @@
+module KeyPairComputeResource
+  extend ActiveSupport::Concern
+
+  included do
+    has_one :key_pair, :foreign_key => :compute_resource_id, :dependent => :destroy
+    after_create :setup_key_pair
+    after_destroy :destroy_key_pair
+    delegate :key_pairs, :to => :client
+  end
+
+  def get_compute_key_pairs
+    return unless capabilities.include?(:key_pair)
+    key_pairs = self.try(:key_pairs)
+    active_key = key_pair
+    return [] if key_pairs.nil? || active_key.nil?
+    compute_resource_key_pair_class = "ComputeResourceKeyPairs::#{provider_friendly_name}".constantize
+    key_pairs.map do |key|
+      compute_resource_key_pair_class.new(key.name, key.fingerprint, active_key.name, active_key.id)
+    end
+  end
+
+  def recreate
+    destroy_key_pair
+    setup_key_pair
+  end
+
+  def delete_key_pair(kay_pair_name)
+    delete_key_from_resource(kay_pair_name)
+  end
+
+  private
+
+  def setup_key_pair
+    key = client.key_pairs.create :name => "foreman-#{id}#{Foreman.uuid}"
+    KeyPair.create! :name => key.name, :compute_resource_id => self.id, :secret => key.private_key
+  rescue => e
+    Foreman::Logging.exception("Failed to generate key pair", e)
+    destroy_key_pair
+    raise
+  end
+
+  def destroy_key_pair
+    return unless key_pair
+    delete_key_from_resource(key_pair.name)
+    key_pair.destroy!
+  end
+
+  def delete_key_from_resource(key_pair_name)
+    logger.info "removing #{provider_friendly_name} key #{key_pair_name}"
```

Comment:
consider adding the compute resource name? e.g. friendly name will only show amazon, but not the region?